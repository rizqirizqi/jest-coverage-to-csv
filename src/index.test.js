import dirTree from 'directory-tree';
import { unlinkSync, readFileSync } from 'fs';
import { resolve as _resolve } from 'path';
import spawn from 'spawn-command';

const EXAMPLE_DIR = 'examples/coverage';

const getInputFile = (dir = EXAMPLE_DIR) => dirTree(dir, { extensions: /\.(json)/ });

const getOutputFile = (dir = EXAMPLE_DIR) => dirTree(dir, { extensions: /\.csv/ });

const clearOutputs = (dir = EXAMPLE_DIR) => {
  const outputFile = getOutputFile(dir);
  outputFile.children.forEach((img) => {
    unlinkSync(img.path);
  });
};

const runCLI = (args = '', cwd = process.cwd()) => {
  const isRelative = cwd[0] !== '/';
  let workingDir = cwd;
  if (isRelative) {
    workingDir = _resolve(__dirname, cwd);
  }

  return new Promise((resolve, reject) => {
    let stdout = '';
    let stderr = '';
    const command = `yarn start ${args}`;
    const child = spawn(command, { cwd: workingDir });

    child.on('error', (error) => {
      reject(error);
    });

    child.stdout.on('data', (data) => {
      stdout += data.toString();
    });

    child.stderr.on('data', (data) => {
      stderr += data.toString();
    });

    child.on('close', () => {
      if (stderr) {
        reject(stderr);
      } else {
        resolve(stdout);
      }
    });
  });
};

describe('Blackbox Test', () => {
  test('jest-coverage-to-csv --help', () => runCLI('--help').then((stdout) => {
    expect(stdout).toMatchSnapshot();
  }));

  describe('Convert Coverage to CSV', () => {
    beforeEach(() => {
      clearOutputs();
    });
    afterEach(() => {
      clearOutputs();
    });
    test('jest-coverage-to-csv examples/coverage/coverage-summary.json  examples/coverage/coverage-summary.csv | it converts the json file', async () => {
      const inputFile = getInputFile();
      let outputFile = getOutputFile();
      expect(inputFile.children.length).toBe(1);
      expect(outputFile.children.length).toBe(0);
      return runCLI(`${EXAMPLE_DIR}/coverage-summary.json ${EXAMPLE_DIR}/coverage-summary.csv`).then((stdout) => {
        outputFile = getOutputFile();
        expect(outputFile.children.length).toBe(1);
        expect(stdout.match(/Conversion to CSV success!/g)).toHaveLength(1);
        const fileContent = readFileSync(`${EXAMPLE_DIR}/coverage-summary.csv`, { encoding: 'utf8' });
        expect(fileContent).toMatchSnapshot();
      });
    });

    test('jest-coverage-to-csv examples/coverage/coverage-summary.json  examples/coverage/coverage-summary.csv --so --po | it converts the json file with option --so --po', async () => {
      const inputFile = getInputFile();
      let outputFile = getOutputFile();
      expect(inputFile.children.length).toBe(1);
      expect(outputFile.children.length).toBe(0);
      return runCLI(`${EXAMPLE_DIR}/coverage-summary.json ${EXAMPLE_DIR}/coverage-summary.csv --so --po`).then((stdout) => {
        outputFile = getOutputFile();
        expect(outputFile.children.length).toBe(1);
        expect(stdout.match(/Conversion to CSV success!/g)).toHaveLength(1);
        const fileContent = readFileSync(`${EXAMPLE_DIR}/coverage-summary.csv`, { encoding: 'utf8' });
        expect(fileContent).toMatchSnapshot();
      });
    });
  });
});
