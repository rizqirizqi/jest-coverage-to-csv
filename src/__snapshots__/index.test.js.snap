// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blackbox Test Convert Coverage to CSV jest-coverage-to-csv examples/coverage/coverage-summary.json examples/coverage/coverage-summary.csv | it converts the json file 1`] = `
"Filename, % Statements, Statements, % Branches, Branches, % Functions, Functions, % Lines, Lines
All Files, 100, 2/2, 80, 4/5, 100, 6/6, 100, 6/6
examples/number.js, 100, 6/6, 80, 4/5, 100, 2/2, 100, 6/6
"
`;

exports[`Blackbox Test Convert Coverage to CSV jest-coverage-to-csv examples/coverage/coverage-summary.json examples/coverage/coverage-summary.csv -f p | it converts the json file with option -f p 1`] = `
"Filename, % Statements, % Branches, % Functions, % Lines
All Files, 100, 80, 100, 100
examples/number.js, 100, 80, 100, 100
"
`;

exports[`Blackbox Test Convert Coverage to CSV jest-coverage-to-csv examples/coverage/coverage-summary.json examples/coverage/coverage-summary.csv -f psl | it converts the json file with option -f psl 1`] = `
"Filename, % Statements, % Lines
All Files, 100, 100
examples/number.js, 100, 100
"
`;

exports[`Blackbox Test jest-coverage-to-csv --help 1`] = `
"$ node ./lib/index.js --help
Convert jest coverage report from coverage-summary to csv
Usage: jest-coverage-to-csv [source] [target] [options]

Source  :  Source path for the coverage-summary json input  [string] [default: \\"./coverage/coverage-summary.json\\"]
Target  :  Target path for the coverage-summary csv output  [string] [default: \\"./coverage/coverage-summary.csv\\"]

Options:
  -f, --filter   Show only specified column. Available filters: p percentages | s statements | b branches | f functions | l lines  [string] [default: \\"\\"]
  -h, --help     Show help  [boolean]
  -v, --version  Show version number  [boolean]

Examples:
  jest-coverage-to-csv
  jest-coverage-to-csv coverage/coverage-summary.json coverage/coverage-summary.csv
  yarn test --coverage --coverageReporters=json-summary && jest-coverage-to-csv
"
`;
